### DISCUSSION TITLE
Error: Prepared statements in promises - INSERT INTO table SET ? --> error INSERT INTO table SET NULL

### DISCUSSION THREAD

**Discussion by TadasV1** [sidorares/node-mysql2#2037] [Q&A]
I have no problem with 
`await connection.execute('INSERT INTO table (dependency_id, title) VALUES (:dependancy_id, :title)', item)`

But i keep getting error when I try to use SET.
```
const item = {
  dependency_id: 11,
  title: 'something'
}

const query = **await** conn.query('INSERT INTO table SET ?', item, function(error, results, fields) {
  if (error) throw error;
})
```

```
ERROR:
Error: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'NULL' at line 1
    at PromiseConnection.query (C:\wamp64\www\hw\node_modules\mysql2\promise.js:93:22)
 ....   
  code: 'ER_PARSE_ERROR',
  errno: 1064,
  sql: 'INSERT INTO food_user SET NULL',
  sqlState: '42000',
  sqlMessage: "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'NULL' at line 1"
}
```

**Comment by sidorares** [comment]
any query you pass to execute must be a valid input for [prepare](https://dev.mysql.com/doc/refman/8.0/en/prepare.html)

in short - you only can pass simple single values ( numbers, strings, dates, buffers ) as a parameters, can't have tables / databases / procedure names as a parameters

The reason this syntax works in `.query()` is because in case of query substitution happen client side

**Comment by TadasV1** [comment]
I'm trying input as provided in example. Since it only shows SET NULL. I have no idea how to debug it. 
But no issues with named syntax.

In my actual case, table has large number of columns. 
No problem with :name syntax. But I need SET syntax since num of fields is dynamic and with VALUES makes it very cumbersome.

**Comment by sidorares** [comment]
You'll need to construct query dynamically in that case ( e.i use `.query` instead of `.execute`).
Alternatively you can pass your data as a JSON blob and handle it in a query or prepared server side statement, in that case you have a single argument ( a string with json data )
